src/drivers/magnetometer/bosch/bmm350/BMM350.cpp

```diff
@@ -0,0 +1,683 @@
+			}
+
+			ScheduleDelayed(100_ms);
+		}
+
+		break;
+
+	case STATE::READ: {
+			// -- start get_compensated_mag_xyz_temp_data
+			int8_t res = 0;
+			uint8_t idx;
+			float out_data[4] = {0.0f};
+			float dut_offset_coeff[3], dut_sensit_coeff[3], dut_tcos[3], dut_tcss[3];
+			float cr_ax_comp_x, cr_ax_comp_y, cr_ax_comp_z;
+
+			res = ReadOutRawData(out_data);
+
+			if (res == 0) {
+				// Apply compensation to temperature reading
+				out_data[3] = (1 + _mag_comp_vals.dut_sensit_coef.t_sens) * out_data[3] +
+					      _mag_comp_vals.dut_offset_coef.t_offs;
+
+				// Store magnetic compensation structure to an array
+				dut_offset_coeff[0] = _mag_comp_vals.dut_offset_coef.offset_x;
+				dut_offset_coeff[1] = _mag_comp_vals.dut_offset_coef.offset_y;
+				dut_offset_coeff[2] = _mag_comp_vals.dut_offset_coef.offset_z;
+
+				dut_sensit_coeff[0] = _mag_comp_vals.dut_sensit_coef.sens_x;
+				dut_sensit_coeff[1] = _mag_comp_vals.dut_sensit_coef.sens_y;
+				dut_sensit_coeff[2] = _mag_comp_vals.dut_sensit_coef.sens_z;
+
+				dut_tcos[0] = _mag_comp_vals.dut_tco.tco_x;
+				dut_tcos[1] = _mag_comp_vals.dut_tco.tco_y;
+				dut_tcos[2] = _mag_comp_vals.dut_tco.tco_z;
+
+				dut_tcss[0] = _mag_comp_vals.dut_tcs.tcs_x;
+				dut_tcss[1] = _mag_comp_vals.dut_tcs.tcs_y;
+				dut_tcss[2] = _mag_comp_vals.dut_tcs.tcs_z;
+
+				for (idx = 0; idx < 3; idx++) {
+					out_data[idx] *= 1 + dut_sensit_coeff[idx];
+					out_data[idx] += dut_offset_coeff[idx];
+					out_data[idx] += dut_tcos[idx] * (out_data[3] - _mag_comp_vals.dut_t0);
+					out_data[idx] /= 1 + dut_tcss[idx] * (out_data[3] - _mag_comp_vals.dut_t0);
+				}
+
+				cr_ax_comp_x = (out_data[0] - _mag_comp_vals.cross_axis.cross_x_y * out_data[1]) /
+					       (1 - _mag_comp_vals.cross_axis.cross_y_x * _mag_comp_vals.cross_axis.cross_x_y);
+				cr_ax_comp_y = (out_data[1] - _mag_comp_vals.cross_axis.cross_y_x * out_data[0]) /
+					       (1 - _mag_comp_vals.cross_axis.cross_y_x * _mag_comp_vals.cross_axis.cross_x_y);
+				cr_ax_comp_z =
+					(out_data[2] +
+					 (out_data[0] *
+					  (_mag_comp_vals.cross_axis.cross_y_x * _mag_comp_vals.cross_axis.cross_z_y -
+					   _mag_comp_vals.cross_axis.cross_z_x) -
+					  out_data[1] *
+					  (_mag_comp_vals.cross_axis.cross_z_y - _mag_comp_vals.cross_axis.cross_x_y *
+					   _mag_comp_vals.cross_axis.cross_z_x)) /
+					 (1 - _mag_comp_vals.cross_axis.cross_y_x * _mag_comp_vals.cross_axis.cross_x_y));
+
+				out_data[0] = cr_ax_comp_x;
+				out_data[1] = cr_ax_comp_y;
+				out_data[2] = cr_ax_comp_z;
+				_px4_mag.set_error_count(perf_event_count(_bad_read_perf) + perf_event_count(_self_test_failed_perf));
+				_px4_mag.set_temperature(out_data[3]);
+				_px4_mag.update(now, cr_ax_comp_x, cr_ax_comp_y, cr_ax_comp_z);
+
+			} else {
+				perf_count(_bad_read_perf);
+			}
+		}
+
+		break;
+	}
+}
+
+bool BMM350::Configure()
+{
+	PX4_DEBUG("Configuring");
+	bool success = true;
+	uint8_t readData = 0;
+
+	// Set pad drive
```

Comment:
Should the pad drive actually be set as early as possible?

Reply:
If I'm not mistaken it should be set before returning to the normal power mode.